test/test_cubed_sphere_circulation.jl

```diff
@@ -0,0 +1,154 @@
+function diagnose_velocities_from_streamfunction(ψ, grid)
+    ψᶠᶠᶜ = Field(Face, Face,   Center, CPU(), grid)
+    uᶠᶜᶜ = Field(Face, Center, Center, CPU(), grid)
+    vᶜᶠᶜ = Field(Center, Face, Center, CPU(), grid)
+
+    for (f, grid_face) in enumerate(grid.faces)
+        Nx, Ny, Nz = size(grid_face)
+
+        ψᶠᶠᶜ_face = ψᶠᶠᶜ.data.faces[f]
+        uᶠᶜᶜ_face = uᶠᶜᶜ.data.faces[f]
+        vᶜᶠᶜ_face = vᶜᶠᶜ.data.faces[f]
+
+        for i in 1:Nx+1, j in 1:Ny+1
+            ψᶠᶠᶜ_face[i, j, 1] = ψ(grid_face.λᶠᶠᵃ[i, j], grid_face.φᶠᶠᵃ[i, j])
+        end
+
+        for i in 1:Nx+1, j in 1:Ny
+            uᶠᶜᶜ_face[i, j, 1] = (ψᶠᶠᶜ_face[i, j, 1] - ψᶠᶠᶜ_face[i, j+1, 1]) / grid.faces[f].Δyᶠᶜᵃ[i, j]
+        end
+
+        for i in 1:Nx, j in 1:Ny+1
+            vᶜᶠᶜ_face[i, j, 1] = (ψᶠᶠᶜ_face[i+1, j, 1] - ψᶠᶠᶜ_face[i, j, 1]) / grid.faces[f].Δxᶜᶠᵃ[i, j]
+        end
+    end
+
+    return uᶠᶜᶜ, vᶜᶠᶜ, ψᶠᶠᶜ
+end
```

Comment:
We should get this working fully abstractly. But I wanted to point out that abstract operations should work on each face individually and so broadcasting could be used:

```julia
for (f, grid_face) in enumerate(grid.faces)
    set!(ψᶠᶠᶜ_face, ψ) # may also need fill_halo_regions!(ψᶠᶠᶜ_face, arch)
    uᶠᶜᶜ_face .= - ∂y(ψᶠᶠᶜ_face)
    vᶜᶠᶜ_face .= + ∂x(ψᶠᶠᶜ_face)
end
```

Then its a short hop to fully broadcasted when we support that in the near future.

Reply:
(And this code also works on the GPU).